source/collectBy.js

```diff
@@ -0,0 +1,44 @@
+import _curry2 from './internal/_curry2';
+import _reduce from './internal/_reduce';
+
+/**
+ * Splits a list into sub-lists, based on the result of calling a String-returning function on each element,
+ * and grouping the results according to values returned.
+ *
+ * @func
+ * @memberOf R
+ * @category List
+ * @sig (a -> String) -> [a] -> [[a]]
```

Comment:
>  based on the result of calling a String-returning function on each element

I am not sure about the constraints of a string returning function, this might be required for `groupBy` but seems to not be strictly needed here.

`collectBy` really looks like `partition` without the need for the returned tuple to be a pair.

```ts
// Tuple length is 2
// eq to partition as fn is a predicate
const [evens, odds] = R.collectBy((n) => n % 2 === 0, [1, 2, 3, 4, 5]);
``` 

```ts
const bySpeed = (speed) => {
  if (speed < 30) { return 0; }
  if (speed < 90) { return 1; }
  return 2;
};

const [slow, average, fast] = R.collectBy(bySpeed, [10, 30, 70, 250]);
```

Probably better to reword

Reply:
I _think_ the string requirement is just a clue for the developer. Related items are grouped under the same key of an object and keys are always strings. The function can indeed return anything but it will be automatically casted into a string.

@CrossEye and I did discuss this already. See https://github.com/ramda/ramda/pull/2715#discussion_r236656183

Reply:
sounds fair, in the end it's more about wording.

I'd personally prefer "A function returning a valid property key".

- [Typescript / PropertyKey](https://github.com/microsoft/TypeScript/blob/master/src/lib/es5.d.ts#L85)
- [Symbols as PropertyKey](https://2ality.com/2014/12/es6-symbols.html#symbols-as-property-keys)

Reply:
I missed this thread when I merged.  @hitmands: I think you're right about the wording.  We should probably change it that way.  Somewhere else we use a signature like

```hs
somefunc :: Index a => (b -> a) -> ?? -> Object
Index == String | Int | Symbol
```

We should probably get that cleaned up and make this a habit for describing functions that return Object keys.  And the description should match.

That can be a separate task, though.